packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart

```diff
@@ -0,0 +1,156 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/widgets.dart';
+
+import 'billing_client_wrapper.dart';
+import 'purchase_wrapper.dart';
+
+/// Abstraction of result of [BillingClient] operation that includes
+/// a [BillingResponse].
+abstract class HasBillingResponse {
+  /// The status of the operation.
+  abstract final BillingResponse responseCode;
+}
+
+/// Utility class that manages a [BillingClient] connection.
+///
+/// Connection is initialized on creation of [BillingClientManager].
+/// If [BillingClient] sends `onBillingServiceDisconnected` event or any
+/// operation returns [BillingResponse.serviceDisconnected], connection is
+/// re-initialized.
+///
+/// [BillingClient] instance is not exposed directly. It can be accessed via
+/// [run] and [runRaw] methods that handle the connection management.
+///
+/// Consider calling [dispose] after the [BillingClient] is no longer needed.
+class BillingClientManager {
+  /// Creates the [BillingClientManager].
+  ///
+  /// Immediately initializes connection to the underlying [BillingClient].
+  BillingClientManager() {
+    _connect();
+  }
+
+  /// Stream of `onPurchasesUpdated` events from the [BillingClient].
+  ///
+  /// This is a broadcast stream, so it can be listened to multiple times.
+  /// A "done" event will be sent after [dispose] is called.
+  late final Stream<PurchasesResultWrapper> purchasesUpdatedStream =
+      _purchasesUpdatedController.stream;
+
+  /// [BillingClient] instance managed by this [BillingClientManager].
+  ///
+  /// In order to access the [BillingClient], consider using [run] and [runRaw]
+  /// methods.
+  @visibleForTesting
+  late final BillingClient client = BillingClient(_onPurchasesUpdated);
+
+  final StreamController<PurchasesResultWrapper> _purchasesUpdatedController =
+      StreamController<PurchasesResultWrapper>.broadcast();
+
+  bool _isConnecting = false;
+  bool _isDisposed = false;
+
+  // Initialized immediately in the constructor, so it's always safe to access.
+  late Future<void> _readyFuture;
+
+  /// Executes the given [block] with access to the underlying [BillingClient].
+  ///
+  /// If necessary, waits for the underlying [BillingClient] to connect.
+  /// If given [block] returns [BillingResponse.serviceDisconnected], it will
+  /// be transparently retried after the connection is restored. Because
+  /// of this, [block] may be called multiple times.
+  ///
+  /// A response with [BillingResponse.serviceDisconnected] may be returned
+  /// in case of [dispose] being called during the operation.
+  ///
+  /// See [runRaw] for operations that do not return a subclass
+  /// of [HasBillingResponse].
+  Future<R> run<R extends HasBillingResponse>(
```

Comment:
I think this method name could be more descriptive. As in it could be `runWhenClientIsReady` or `runOnClientAvailable`. I think this would make it easier to understand when this method is being used in the code. Same goes for the method `runRaw`.

Reply:
I also agree that `run`/`runRaw` is ambiguous. I don't think that `runWhenClientIsReady`, `runOnClientAvailable` would be good though. Main idea of `BillingClientManager` is to handle `BillingClient` connection transparently, so its user doesn't have to worry about client readiness/availability/retries. I suggest `withClient` and `withClientNonRetryable` instead.

Reply:
How about `runWithClient`?

Reply:
Sure, done.